### ISSUE TITLE
Problem with Nematus

### ISSUE THREAD

**Issue by slavikkom** [mt2017-tartu-shared-task/nmt-system-C#7]
How to run Nematus nmt.py properly? It has parameters --datasets (paths to paralell corpora - it should be train.et, train.en - or after some preprocessing (e.g. tc-tok-train)), and --vocabulary - which vocabulary it should be - generated with OpenNMT-py?

**Comment by MaksymDel** [comment]
Hi,

you should provide preprocessed (including BPE) datasets to the training script. As for vocabulary, it is called dictionary in Nematus, and you can build it following way:
`# build network dictionary
$nematus/data/build_dictionary.py data/corpus.bpe.$SRC data/corpus.bpe.$TRG`

To run Nematus really properly, refer to following example: 
https://github.com/rsennrich/wmt16-scripts/tree/master/sample

It has train.sh file, which has device parameter (set to cpu if you train on cpu) and runs training script which is defined in config.py file.
The config.py file also shows a way to specify other parameters related to training.  

Finally, translate.sh shows how to translate when you have some model trained, and wants you to specify paths to test datasets.

**Comment by slavikkom** [comment]
I successfully build dictionaries, after that I run training with all default parameters - is it ok to use all default parameters (I specified only paths to training sets, dev sets and dictionaries)? Furthermore, when running on cpu, all jobs that specify memory more than 50gb are blocked. I specify precisely 50gb, and after 2 hours script works well. Can it be that after some time it stops, or if script did't stop in first hour, then everything will work?

**Comment by MaksymDel** [comment]
It is OK to use default parameters if you do not explicitly need to modify some of them. Just make sure you know what they are and can list all of them later.

As for your script, as soon as you run it successfully, we assume it works. Nematus is stable enough, and I do not expect memory leaks or some other types of crucial probles like this.

However, I strongly recommend you to check whether your script is running as well as validation and train errors behavior. You can also try to translate with some intermediate checkpoints without interrupting the model.

**Comment by slavikkom** [comment]
How much gpu faster than cpu? I wonder that after 16 hours there is no intermediate results (such as first epoch or something like that). Nematus saves checkpoints in the same way as OpenNMT-py?

**Comment by slavikkom** [comment]
After some time, script stops because of lack of memory. Can I somehow increase memory size for cpu (use 2 nodes or something like that)?

**Comment by MaksymDel** [comment]
GPU is usually 5 to 50 times faster.

How much memory do you use now?

**Comment by slavikkom** [comment]
I used 55gb, when I tried to allocate more, I receive message: "batch job submission error. Requested node configuration is not available".

**Comment by MaksymDel** [comment]
Please attach your parameters file as well as running script

**Comment by slavikkom** [comment]
#!/bin/bash

#The name of the job is test_job
#SBATCH -J train_NMT

#The job requires 1 compute node
#SBATCH -N 1

#The job requires 1 task per node
#SBATCH --ntasks-per-node=1

#The maximum walltime of the job is a 8 days
#SBATCH -t 192:00:00

#SBATCH --mem=50G

#Leave this here if you need a GPU for your job
#SBATCH --partition=long

#The module with all the NMT / deep learning packages
module load python-2.7.13

device=cpu

THEANO_FLAGS=mode=FAST_RUN,floatX=float32,device=$device,on_unused_input=warn python nmt.py --datasets ../data/corpus.bpe.et ../data/corpus.bpe.en --dictionaries ../data/corpus.bpe.et.json ../data/corpus.bpe.en.json --model ./model/model.npz

**Comment by slavikkom** [comment]
[model.txt](https://github.com/mt2017-tartu-shared-task/nmt-system-C/files/1537524/model.txt)

**Comment by MaksymDel** [comment]
As about logging and saving frequency, check *Freq parameters in your model.txt.

As for memory, usually smaller vocab and/or batch size helps.

**Comment by slavikkom** [comment]
How much iterations are in one epoche? How can I precalculate it (for example, in OpenMNT is was 281599 ).

**Comment by MaksymDel** [comment]
To perform one iteration means to process one minibatch. So the number of iterations equals the number of mini-batches needed to process all examples in dataset.

**Comment by slavikkom** [comment]
Nematus stores model after each 30000 iterations, can I continue training not after some epoch, but from some of intermediate stored model (for example, from model after 210000 iterations in first epoche) - to decrease time for training one epoche (by decreasing maximum number of minibatches parameter).

**Comment by MaksymDel** [comment]
You can resume training from model checkpoints. 
You can not resume training from the point you do not have a model checkpoint at.

"saveFreq": 30.000 iterations means creating model checkpoint after each 30.000 mini batches
Your dataset is ~600.000 data points. Divide it by your minibatch size, and you will know how many iterations (mini batches) it takes to perform one epoch. 

> "maximum number of mini batches parameter"

You can not change this explicitly. You can only increase size of your minibatch, and thus you will have to pass fewer mini-batches to cover all points in your dataset (do one epoch)

Did I answer your question?
And at the end, what do you want to achieve?

**Comment by slavikkom** [comment]
Yes. Currently I am training model at Google cloud gpu, speed is approximately 120000 iterations per day, and I was wondering whether this speed is enough. In the end I am planning to translate test set with ensemble decoding (and tunning beam size).

**Comment by MaksymDel** [comment]
I would suggest you study a beam size first (since you can do it quickly CPU and can write about it in your report in any case, but it is up to you at the end). 

As for GCG, I do not know from where your credits come from, but in the report, you have a corresponding section that is about what you did not try but wanted (because of GPUs for example).

Just in case, you showed me you can get topN output from OpenNMT, so you could also (for example) manually study these additional sentences, and tell, whether there is something better then top1 output, or not. Then compare it to metrics from the visualization tool we used, etc. There is a big room what you can do even without GPUs.

 Also, I would suggest to concentrate you your report more.
